### ISSUE TITLE
 pgr_vidstodmatrix

### ISSUE THREAD

**Issue by cvvergara** [pgRouting/pgrouting#368]
@yobiSource @woodbri 
A space to talk about  pgr_vidstodmatrix
Discussion Started on issue #205

**Comment by cvvergara** [comment]
Transcribing https://github.com/pgRouting/pgrouting/issues/205#issuecomment-122645715
"
pgr_vidstodmatrix did not exist in the V2.0 release, it was added later.

I have slightly customized pgr_vidstodmatrix and made it smaller for my purposes.
But the return value is still a Dmatrix, just as an array instead of a record with arrays.
(I have to call the function a thousand times with up to 500x500 route variants. This takes hours and eats my RAM.)

I've compiled pgrouting-2.1.0-alpha1 and both versions exist.
old pgr_vidstodmatrix(integer[], postgis.geometry[], text, double precision)
RETURNS record AS dmatrix double precision[], ids integer[]

new pgr_vidstodmatrix(text, integer[], boolean, boolean, boolean)
RETURNS double precision[] (As my modified version of the old)

I have run both functions with the sample data and both return the same Dmatrix.
Exactly what I need, the SQL function in C for faster execution.
The rest is pgr_tsp job.

Perhaps @woodbri can explain the new function or add missing doc.
(I do not know what "dir boolean" does)

I have to prepare my server, for testing the large amounts of OSM data.

evolve pgr_kdijkstraCost and pgr_kdijkstraPath is surely the best option but I can not wait. ^^
"

**Comment by woodbri** [comment]
dir boolean is the flag to create a directed graph instead of an undirected graph.

In general I think there can be some rationalization of the functions. It has been I while since I last looked at these, but we need the ability to create both symmetric and asymmetric distance matrices. The original TSP function requires a symmetric matrix, but  Dave Potts did some work to support add support for asymmetric TSP. I'm not sure if that work got checked in or it is in a fork somewhere.

The original plpgsql function vidstodmatrix was written before the C code as a prototype. I'm fine with that implementation going away in favor of a C based function especially if the C function is faster. The idea for the C function should be along the lines of build one big graph the includes all the nodes then solve that multiple times without rebuilding it multiple times, which is what I think is happening.

We could include (in fact I think I already have included) plpgsql code to convert records to a dmatrix and convert a dmatrix to records. We need to keep function simple and focused to a single task and then write conversion functions to transform the data to new structures as needed by other commands. This make the code more atomic and modular and easier to test and debug. Also changes are not likely to break other code or have bad side effects because the testing of atomic features is easier to validate them.

**Comment by cvvergara** [comment]
Some of the reasons for not including pgr_vidsToDMatrix officialy can be found in this
[comment](https://github.com/pgRouting/pgrouting/issues/205#issuecomment-122601644)
It still does the [renumbering](https://github.com/pgRouting/pgrouting/blob/develop/src/kdijkstra/src/k_targets_sp.c#L1101) in the C code. (so if you have 2 vertices id 1 and an edge id 100000 it stills creates an 100000 size array for those 2 vertices) 
and issues of its own: like this 
 - [line 98 FIXME](https://github.com/pgRouting/pgrouting/blob/develop/src/kdijkstra/src/k_targets_boost_wrapper.cpp#L98)
 - [line 104 FIXME](https://github.com/pgRouting/pgrouting/blob/develop/src/kdijkstra/src/k_targets_boost_wrapper.cpp#L104)
 - [line 306 FIXME](https://github.com/pgRouting/pgrouting/blob/develop/src/kdijkstra/src/k_targets_boost_wrapper.cpp#L306)
 -  [line 312 FIXME](https://github.com/pgRouting/pgrouting/blob/develop/src/kdijkstra/src/k_targets_boost_wrapper.cpp#L312)

**Comment by woodbri** [comment]
I'm not sure we would want to apply that filter to all functions or we would have to remove a lot of others also. On the other hand, we might want to document a restriction and leave the ticket open to remove those restrictions now or in the future.

**Comment by cvvergara** [comment]
By "that filter" I think you mean the functions with bugs.
The function pgr_vidstodmatrix exist, it is useful and being used, it didn't exist in V2.0, and its kept unofficially.
Functions that existed in V2.1 can not be removed, they have to stay an live with the bugs, and wait for their turn to be re factored. But if we officially start accepting officially functions that have bugs this will be a never ending story.

**Comment by woodbri** [comment]
V2.1 does allow for new functions to be added, the only constraint is that we should not break any officially supported function from V2.0.

Regarding bug versus design limitation, is a little more subjective. pgr_vidstodmatrix works as designed and if we document the restrictions then it should be fine. This is not to say that it could have been designed better or with less restrictions. I will not have time to rework it for this release, so the PSC can decide if the value out weighs the risk of adding it.

**Comment by yobiSource** [comment]
I had some time to test the functions with my server and OSM data.
Here are the data for 500 times calling pgr_vidstodmatrix, with around 4-497 ids and 14 - 10139 edges per call. (with TSP)

version | time | RAM | resolved
------------ | ------------- | ------------- | -------------
SQL | 2194s | evenly up to 3.9 GB ! | 482
C | 1500s | unsteady 200-250 MB | 500

RAM is only released when the connection to PostgreSQL is closed, even if the function has been completed!

* 78 have the same route lengths
* 394 have a difference of less than 10% (shorter and longer)
* 10 with more than 10% (one 56%)

The problem lies in the DMatrix with slightly different distances. (Thus TSP reacts differently)
I think the numbers are rounded. (stupid MS Excel)

SQL ```onetomany_dijkstra_dist```  | C ```manytomany_dijkstra_dmatrix```
------------ | -------------
746,193354142594 | 746,193354142593
727,073600194356 | 727,073600194357
365,331599679096 | 365,331599679095
645,161303384438 | 645,161303384439

So the SQL version takes longer and is an OOM killer.

**Comment by cvvergara** [comment]
Wow @yobiSource I am impressed, 
Although pgr_dijkstra you can use it to generate a table instead of a matrix, you can have a table representation of a cost matrix: you need to set the array data for the source, and the array data for the target to be the same like in this example:
```
select * from pgr_dijkstra( <sql> , array[1,2,3,4], array[1,2,3,4]);
```
Would you do a similar test with this idea?

P.D
I like tests.
You might like [statsQuery](https://gist.github.com/cvvergara/dfe41b7d566536102df3) function I used to make some statistical analysis of running times for the pgRouting functions. (I only used it in the new functionality functions) to have a good degree of freedom  I ran it with  n=30 (number of times the test run to get an average time) The m, basically is to control the number of edges in the query.

**Comment by yobiSource** [comment]
Cool, I had not yet found time to look at the new features.

A quick test with 3 IDs shows 2 routes.
For a Symmetric Matrix one route is missing, was none found?

statsQuery looks helpful.
I think I will replace ```LIMIT``` with ```bbox2d``` so that the results remain comparable with the first Tests.

If none exists, I can make over the weekend a function that converts  the table to an array for pgr_TSP.

**Comment by cvvergara** [comment]
- About  quick test:
Default is on directed graph. So probably it is, but I am only guessing as I don't know your data, the query, and the results.
Say that from 2 to 3 is not in the table, then:
```
 select * from pgr_dijkstra( <sql> ,2, 3);
```
should  be empty
to try undirected:
```
 select * from pgr_dijkstra( <sql> , <source>, <target>, directed:=false);
```

**Comment by cvvergara** [comment]
- About statsQuery, Modify as you need. (that is what I do, its being changing depending on what I want to measure)

**Comment by yobiSource** [comment]
```directed:=false``` that's it now I have all variants

**Comment by cvvergara** [comment]
@yobiSource well, as I teld you it was a guess, I can't do more without data.

**Comment by yobiSource** [comment]
I've created a function that converts the data from pgr_dijkstra to an array (dmatrix).
The only problem was the poor performance.
The following loop took 26 seconds out of 32 seconds (over 100000 loops):

```sql
for i in 1 .. array_length(aggcost,1) loop 
	 dmatrix[startids[i]][endids[i]] := aggcost[i];
end loop;
```
PostgreSQL is very slow with large arrays thus I have switched to temporary tables
and it became much faster, just **1** second.
I still have to check the correctness of the dmatrix but it looks good.
For the future it would be better if pgr_tsp works directly with the data of pgr_dijkstra.

In general, the new pgr_dijkstra is clearly better and more flexible, so I can improve my own functions. :smile:

**Comment by yobiSource** [comment]
Tests were successful.

version | time | RAM | resolved
------------ | ------------- | ------------- | -------------
SQL V2 | 440s | unsteady 400-600 MB | 482

function for converting TSP ready:
https://gist.github.com/yobiSource/44b0475b2d050a7b676e#file-new-vidstodmatrix

function call:
```sql
DROP TABLE IF EXISTS tmp_vidscost;

CREATE TEMP TABLE tmp_vidscost ON COMMIT DROP AS
SELECT -start_v AS start_v, -end_v AS end_v, agg_cost
FROM
  (SELECT *
   FROM pgr_dijkstra('SELECT id::integer, source::integer, target::integer, cost::float8 FROM t_osm_ways_n where geom && ' || 
   quote_literal(bbox::text) || '::box2d'::text, vertex_ids, vertex_ids, directed:=FALSE)
   WHERE edge = -1) AS sel;

SELECT f_vidstodmatrix2(vertex_ids::bigint[])
```

**Comment by cvvergara** [comment]
@yobiSource 
One of the new features in pgr_dijkstra,  is that it can accept
  *  int, smallint, bigint in id's
  *  int, smallint, bigint, real, float in costs

Can you try:
```
DROP TABLE IF EXISTS tmp_vidscost;

CREATE TEMP TABLE tmp_vidscost ON COMMIT DROP AS
SELECT -start_v AS start_v, -end_v AS end_v, agg_cost
FROM
  (SELECT *
   FROM pgr_dijkstra('SELECT id, source, target, cost FROM t_osm_ways_n where geom && ' || 
   quote_literal(bbox::text) || '::box2d'::text, vertex_ids, vertex_ids, directed:=FALSE)
   WHERE edge = -1) AS sel;

SELECT f_vidstodmatrix2(vertex_ids::bigint[])
```

**Comment by yobiSource** [comment]
works without problems

**Comment by cvvergara** [comment]
I am trying to understand how this function works:
As the example in the [documentation](http://docs.pgrouting.org/2.1/src/common/doc/convenience/vids_to_dmatrix.html) doesn't show the results of he matrix, I am using a smaller graph extracted from the sample data of the documentation
```
SELECT * into petit_graph FROM edge_table WHERE id < 3;
SELECT pgr_createVerticesTable('petit_graph');
```

so petit_graph has: (and the_geom that has the edge geometry not shown here'
```
SELECT id , source , target , cost , reverse_cost , x1 , y1 , x2 , y2
        FROM petit_graph;
 id | source | target | cost | reverse_cost | x1 | y1 | x2 | y2 
----+--------+--------+------+--------------+----+----+----+----
  1 |      1 |      2 |    1 |            1 |  2 |  0 |  2 |  1
  2 |      2 |      3 |   -1 |            1 |  2 |  1 |  3 |  1
(2 rows)
```

with  pgr_createVerticesTable we also have the corresponding vertices table;
```
SELECT id, the_geom FROM petit_graph_vertices_pgr;
 id |                  the_geom                  
----+--------------------------------------------
  1 | 01010000000000000000000040000000000000F03F
  2 | 01010000000000000000000840000000000000F03F
  3 | 010100000000000000000000400000000000000000
(3 rows)
```

Based on this data and using reverse_cost column:
 - if the graph is directed the matrix would look like:
```
       1     2      3
---  ----  ----   ----
1     0     1     inf
2     1     0     inf
3     1     1     0
```

 - if the graph is undirected the matrix would look like:
```
        1     2    3
---  ----  ----   ----
1     0     1      2
2     1     0      1
3     2     1     0
```

Following the documentation:

- vids:	integer[] - An array of vertex_id.
  - array[1,2,3]  

  - this values be obtained with a query like this
 ```
SELECT array_agg(id) FROM petit_graph_vertices_pgr;
 array_agg 
-----------
 {1,2,3}
(1 row)
```


- pnts:	geometry[] - An array of point geometries that approximates the extents of the vertex_id.
  - array[ 01010000000000000000000040000000000000F03F,
  01010000000000000000000840000000000000F03F,
  010100000000000000000000400000000000000000]

  - This values can be obtained with a query like this
```
SELECT array_agg(the_geom) FROM petit_graph_vertices_pgr;
                                                             array_agg                                                              
------------------------------------------------------------------------------------------------------------------------------------
 {010100000000000000000000400000000000000000:01010000000000000000000040000000000000F03F:01010000000000000000000840000000000000F03F}
(1 row)
```

- edges:	text - The edge table to be used for the conversion.
  - edges  = 'petit_graph'

- tol:	float8 - The amount to expand the BBOX extents of pnts when building the graph.
 - 0
 - As for this example the original data is being used, 0 is good

So lets try it !!!!:

```
WITH vids 
AS (SELECT array_agg(id) AS vids FROM petit_graph_vertices_pgr),
geoms as (SELECT array_agg(the_geom) AS pants FROM petit_graph_vertices_pgr)
SELECT * FROM pgr_vidsToDMatrix(
vids, geoms, 'petit_graph', 0);
```

hmm,  maybe I missed Something

**Comment by cvvergara** [comment]
From the last post
How do I make this query work?
```
WITH vids 
AS (SELECT array_agg(id) AS vids FROM petit_graph_vertices_pgr),
geoms as (SELECT array_agg(the_geom) AS pants FROM petit_graph_vertices_pgr)
SELECT * FROM pgr_vidsToDMatrix(
vids, geoms, 'petit_graph', 0);
```

**Comment by robe2** [comment]
try either:

```
with vids AS (select array_agg(id) as vids from petit_graph_vertices_pgr),
geoms as (select array_agg(the_geom) as pnts from petit_graph_vertices_pgr)
select * from pgr_vidsToDMatrix (
(select vids::integer[] from vids) , (select pnts from geoms), 'petit_graph'::text,
0.0::float);

```

or

```

with vids AS (select array_agg(id::integer) as vids from petit_graph_vertices_pgr),
geoms as (select array_agg(the_geom) as pnts from petit_graph_vertices_pgr)
select * from pgr_vidsToDMatrix (
(select vids from vids) , (select pnts from geoms), 'petit_graph'::text,
0.0::float);

```

**Comment by cvvergara** [comment]
Thanks, @robe2 
I tried both,
but unfortunately it doesn’t stop, at least after 5 minutes I had to stop it.

**Comment by cvvergara** [comment]
This function(s) are for doing some pre-process before calling pgr_TSP

There is documentation of 2 different `pgr_vidsToDmatrix`:
- http://docs.pgrouting.org/2.1/src/common/doc/convenience/vids_to_dmatrix.html
- http://docs.pgrouting.org/2.1/src/common/doc/convenience/vids_to_dmatrix2.html#pgr-vids-to-dmatrix2

- This proposal doesn't support bigint for id's (issue #140 opened) by @woodbri 

- Considering that both use `pgr_kDikjstraCost`:
  -  In this [description](http://docs.pgrouting.org/2.1/src/common/doc/convenience/vids_to_dmatrix2.html#description) the developer noticed that `pgr_kDijkstraCost` has failures
  -  `pgr_kdijkstraCost` is going to be deprecated in V2.2 in favor of  `pgr_dijkstraCost`
  -  `pgr_kdijkstraCost` will be giving the correct results during the V2.x (x >= 2)
  -  `pgr_kdijkstraCost` only works for 1 to many

One thing that needs to be noticed, about [boost's TSP ](http://www.boost.org/doc/libs/1_59_0/libs/graph/doc/metric_tsp_approx.html)
- Doesn't not work with negative values.
- Works only on undirected graph
- The matrix must obey the triangle inequality.
- There is a trick on how to set the matrix when used for directed graph, coded by @dapotts in this gist [here](https://gist.github.com/dapotts/6133259) that obeys the triangle inequality

The current implementation of TSP in pgrRouting doesn't use boost's TSP, 
So here are some questions:
- It does not accept negative values as distances [here](https://github.com/pgRouting/pgrouting/blob/master/src/tsp/src/tsplib.c#L213), so why  `pgr_vidsToDmatrix` return's negative values?
- does the current TSP must also obey  the triangle inequality?

So I think it would be a nice improvement to this proposal to:
- make it work for ANY-INTEGER on the id's, and ANY-NUMERICAL for the costs
- return EMPTY MATRIX If the triangle inequality is not obeyed.
- instead of retuning -1 when there is no path, return an "infinite" value which can be the maximum value accepted by the datatype.
- instead of using pgr_kdijkstraCost, use pgr_dijkstraCost many to many functionality.

**Comment by woodbri** [comment]
* does the current TSP must also obey the triangle inequality?
     - Yes I believe that it does.
* It does not accept negative values as distances here, so why pgr_vidsToDmatrix return's negative values?
    - because it is expensive to compute the results and if a route fails we set the result to -1. The expectation is the the user will check that the results are valid before using them and if their are -1 in the results that they know which pairs failed so they can report that to the user and decide how they want to proceed.

**Comment by cvvergara** [comment]
The example in the documentation directly uses the results in pgr_tsp, and no mention of that expectation is in the documentation.

**Comment by woodbri** [comment]
So the doc is inadequate and maybe needs and additional example that returns -1 to show it is possible. The idea is that pgr_vidsToDmatrix might be used  for functions other than pgr_tsp, like VRP problems or user written code.

**Comment by yobiSource** [comment]
@cvvergara The queries of @robe2 work for me. Here is another version:
```sql
SELECT * FROM pgr_vidsToDMatrix(
	(SELECT array_agg(id::integer) AS vids FROM petit_graph_vertices_pgr), 
	(SELECT array_agg(the_geom) AS pants FROM petit_graph_vertices_pgr),
	'petit_graph'::text, 
	0::double precision
	);
```
In my opinion, both ```pgr_vidsToDmatrix``` functions are deprecated.
With the new ```pgr_dijkstra``` and my [vidstodmatrix](https://gist.github.com/yobiSource/44b0475b2d050a7b676e#file-new-vidstodmatrix) from above everything is much [faster](https://github.com/pgRouting/pgrouting/issues/368#issuecomment-137967101).

But the best would be if the ```pgr_TSP``` process data directly from ```pgr_dijkstra```, like:
```sql
SELECT * FROM pgr_tsp( pgr_dijkstra(edges_sql, start_vid, end_vid) )
```

**Comment by cvvergara** [comment]
@yobiSource 
I can bet you made some time analysis like the above.
So what about the following idea (because we are not modifying TSP) for V2.2
As we can not "deprecate" something that is proposed, (the proposal works for discussion testing, analysis, etc Just like we are doing now)
Why don't you propose pgr_convertToMatrix or something similar that can be used like this:
```
 pgr_convertToMatrix( pgr_dijkstraCost(<parameters of pgr_dijkstraCost>) )
```
So that could be the input to TSP

Things that you should care about:
- [pgr_dijkstraCost](http://docs.pgrouting.org/dev/src/dijkstra/doc/dijkstraCost.html) is proposed.
- But works for BIGINT as id's (as we are moving everything to also handle BIGINT)
- Basically is this is pgr_dijkstraCost and you can simulate it using this query
```
  SELECT start_vid, end_vid, agg_cost 
           FROM pgr_dijkstra(<parameters of pgr_dijkstra>) 
           WHERE edge = -1;
```
  - except that when its from the same source and target it will return a record (v,v,0)
  - infinity values (when there is no path) are not returned, so if the combination (u,v, val) is not there it is because there is no path from u to v. 

You can use the same branch: nodeNet-dev-2.2 like the last time, I can do the merge.

**Comment by cvvergara** [comment]
@woodbri 
There are 2 of them:
http://docs.pgrouting.org/dev/src/convinience/doc/pgr_vidsToDMatrix.html
http://docs.pgrouting.org/dev/src/convinience/doc/pgr_vidsToDMatrix2.html

One of them uses plpsql the other uses internally pgr_kdijkstraCost code.
kdijstraCost is now wrapped by pgr_dijkstraCost

Some differences to consider:
pgr_dijkstraCost does not return an answer when:
  - (u,u) because cost is allways 0
  - when there is no path (u,v) the cost is infinity

**Comment by cvvergara** [comment]
I kept the code of under kdijstra directory just for this function to work.

**Comment by cvvergara** [comment]
Functions that return a distance matrix. A distance matrix is used in TSP as input

* https://github.com/pgRouting/pgrouting/blob/master/src/convinience/sql/pgrouting_dmatrix_tools.sql#L66
* https://github.com/pgRouting/pgrouting/blob/6657190d687b3f7b812feac12afbcea73fc79f5a/src/tsp/sql/routing_tsp_wrappers.sql#L37
* https://github.com/pgRouting/pgrouting/blob/master/src/convinience/sql/pgrouting_dmatrix_tools.sql#L23
* https://github.com/pgRouting/pgrouting/blob/master/src/convinience/sql/pgr_vidsToDMatrix_dir.sql

As I see: I would like this naming:
pgr_euclideanDMatrix instead of  _pgr_makeDistanceMatrix
pgr_dijkstraDMatrix instead of pgr_vidsToDMatrix (both of them)
pgr_withPointsDMatrix instead of pgr_pointsToDMatrix
----- for the future example---
pgr_astarDMatrix
etc...

The name implies the inside functionality. (or so I think)
and also, the parameters needed would be similar to the family of functions to which they belong.

As they are proposed, no backwards compatibility is necessary.

>  This is a proposed function
    Is not officially in the release.
    Name could change.
    Signature could change.
    Needs testing.
    Functionality could change.

**Comment by cvvergara** [comment]
@yobiSource has point here:
https://github.com/pgRouting/pgrouting/issues/368#issuecomment-136415210
> PostgreSQL is very slow with large arrays thus I have switched to temporary tables
and it became much faster, just 1 second.

for some reason, (maybe because internally the C/C++ code does use arrays) the data is passed in form of a matrix.
So instead of giving back the results as a matrix, array[][]
give them as set of records of
start_vid, end_vid, agg_cost
Note: its the same kind of output returned on pgr_floydWarshall & pgr_johnson

non existing records paths (u, v) don't have a row -> agg_cost is infinity
non existing records paths (v, v) don't have a row -> agg_cost is 0

Note: agg_cost to separate the fact that cost is used on an edge.

That would require TSP to have as input that kind of outputs.


The Benefits:
if the user needs to make some adjustments, maybe because the DMatrix was generated from a directed graph (unsymmetrical matrix) the user can decide how to convert it to symmetrical.
The user can keep a permanent table and add this results for future reference, (maybe not calculate everything but use previous calculated data)
 
In this case,
pgr_euclideanDMatrix(this return record)   & keep _pgr_makeDistanceMatrix(this return matrix)
pgr_dijkstraDMatrix(this return record) & keep  pgr_vidsToDMatrix (one of them)(this return matrix)
pgr_withPointsDMatrix(this return record) & keep pgr_pointsToDMatrix(this return matrix)

because it would mean that pgr_TSP, would need the backwards compatibility.

**Comment by cvvergara** [comment]
For example:
```
SELECT * FROM pgr_dijkstraDMatrix(
    'SELECT id, source, target, cost, reverse_cost, x1, y1, x2, y2 FROM edge_table',
    array[2,5,7,9]);
 start_vid | end_vid | agg_cost 
-----------+---------+----------
         2 |       5 |        1
         2 |       7 |        3
         2 |       9 |        3
         5 |       2 |        1
         5 |       7 |        2
         5 |       9 |        2
         7 |       2 |        3
         7 |       5 |        2
         7 |       9 |        4
         9 |       2 |        3
         9 |       5 |        2
         9 |       7 |        4
(12 rows)
```

**Comment by cvvergara** [comment]
Other example:
```
 SELECT * FROM pgr_withPointsDMatrix(
        'SELECT id, source, target, cost, reverse_cost FROM edge_table ORDER BY id',
        'SELECT pid, edge_id, fraction, side from pointsOfInterest',
        array[-1, 3, 5, 6, -6]);
 start_pid | end_pid | agg_cost 
-----------+---------+----------
        -6 |      -1 |      1.3
        -6 |       3 |      4.3
        -6 |       5 |      0.3
        -6 |       6 |      1.3
        -1 |      -6 |      1.3
        -1 |       3 |      5.6
        -1 |       5 |      1.6
        -1 |       6 |      2.6
         3 |      -6 |      1.7
         3 |      -1 |      1.6
         3 |       5 |        2
         3 |       6 |        1
         5 |      -6 |      0.3
         5 |      -1 |      1.6
         5 |       3 |        4
         5 |       6 |        1
         6 |      -6 |      1.3
         6 |      -1 |      2.6
         6 |       3 |        3
         6 |       5 |        1
(20 rows)
```

**Comment by cvvergara** [comment]
Sometimes the array of id's is a result of a query, so remove duplicates if needed:
```
SELECT * FROM pgr_dijkstraDMatrix(
    'SELECT id, source, target, cost, reverse_cost FROM edge_table',
    array[2,5,7,5,7]);
 start_vid | end_vid | agg_cost 
-----------+---------+----------
         2 |       5 |        1
         2 |       7 |        3
         5 |       2 |        1
         5 |       7 |        2
         7 |       2 |        3
         7 |       5 |        2
(6 rows)
```

**Comment by cvvergara** [comment]
For the future example:
```
 SELECT * FROM pgr_astarDMatrix(
    'SELECT id, source, target, cost, reverse_cost, x1, y1, x2, y2 FROM edge_table',
    array[2,5,7,9], heuristic:=4);
 start_vid | end_vid | agg_cost 
-----------+---------+----------
         2 |       5 |        1
         2 |       7 |        3
         2 |       9 |        3
         5 |       2 |        1
         5 |       7 |        2
         5 |       9 |        2
         7 |       2 |        3
         7 |       5 |        2
         7 |       9 |        4
         9 |       2 |        3
         9 |       5 |        2
         9 |       7 |        4
(12 rows)
```

**Comment by woodbri** [comment]
+1, these all look like reason changes and I support them.

**Comment by cvvergara** [comment]
Thanks...
working branch:
https://github.com/cvvergara/pgrouting/tree/newFunction/DMatrix

**Comment by cvvergara** [comment]
For the pgr_eucledianDMatrix...
I can see 2 flavours:
for example, when the data comes from edge_table_vertices_pgr
```
select * from pgr_eucledianDMatrix('select id, the_geom from edge_table_vertices_pgr')
```

for example, when the data comes from edge_table (maybe because the edge_table_vertices_pgr does not exist)
```
select * from pgr_eucledianDMatrix(
'SELECT source AS id, x1 AS x, y1 AS y from edge_table
UNION
SELECT target AS id, x1 as x, y1 as y from edge_table'
) 

select * from pgr_eucledianDMatrix(
'SELECT id, ST_X(the_geom) AS x,  ST_Y(the_geom) AS y from edge_table_pgr'
) 
```
duplicated id's would be removed.
(haven't made a complete example yet)

**Comment by woodbri** [comment]
I think a third example might be useful where we take a string like "id,lat,lon;id,lat,lon;..." ie: something you might get from a web page and show how to unpack that into a query like:
```
select * from pgr_eucledianDMatrix($$
  with pt as (
    select * from regexp_split_to_table('1,10,20;2,11,21;3,12,22', ';') pt
  ),
  array1 as (
    select regexp_split_to_array(pt, ',') as arry from pt
  )
  select arry[1]::integer as id, arry[2]::float8 as y, arry[3]::float8 as x) from array1
$$);
```
The idea being that you don't *have* to have a table if your data is coming from a non table source.

**Comment by cvvergara** [comment]
maybe I didnt explain my self...
pgr_eucledianDMatrix(vertices_sql) RETURNS set of start_vid, end_vid, add_cost
where vertices sql has
id, geometry columns

or (and if possible)

pgr_eucledianDMatrix(vertices_sql) RETURNS set of start_vid, end_vid, add_cost
where vertices sql has
id, x ,y columns

**Comment by cvvergara** [comment]
For example:
```
SELECT * from pgr_eucledianDMatrix('
          SELECT id, the_geom FROM edge_table_vertices_pgr 
          WHERE id IN (1,5,6,7)');
 start_vid | end_vid |     agg_cost     
-----------+---------+------------------
         1 |       5 |                2
         1 |       6 | 2.23606797749979
         1 |       7 | 2.82842712474619
         5 |       1 |                2
         5 |       6 |                1
         5 |       7 |                2
         6 |       1 | 2.23606797749979
         6 |       5 |                1
         6 |       7 |                3
         7 |       1 | 2.82842712474619
         7 |       5 |                2
         7 |       6 |                3
(12 rows)
```

**Comment by cvvergara** [comment]
Example when vertices_sql has (id,x,y) columns
```
SELECT * from pgr_eucledianDMatrix('SELECT id, ST_X(the_geom) AS x, ST_Y(the_geom) AS y FROM edge_table_vertices_pgr WHERE id IN (1,5,6,7)');
 start_vid | end_vid |     agg_cost     
-----------+---------+------------------
         1 |       5 |                2
         1 |       6 | 2.23606797749979
         1 |       7 | 2.82842712474619
         5 |       1 |                2
         5 |       6 |                1
         5 |       7 |                2
         6 |       1 | 2.23606797749979
         6 |       5 |                1
         6 |       7 |                3
         7 |       1 | 2.82842712474619
         7 |       5 |                2
         7 |       6 |                3
(12 rows)
```

**Comment by cvvergara** [comment]
hmm I still need to remove duplicates
```
 select * from pgr_eucledianDMatrix(
'SELECT source AS id, x1 AS x, y1 AS y from edge_table WHERE source IN (1,5,6,7)
UNION
SELECT target AS id, x1 as x, y1 as y from edge_table WHERE target IN (1,5,6,7)'
) ;
 start_vid | end_vid |     agg_cost     
-----------+---------+------------------
         1 |       5 |                2
         1 |       5 | 2.23606797749979
         1 |       5 |                1
         1 |       6 |  1.4142135623731
         1 |       6 | 2.23606797749979
         1 |       6 |                2
         1 |       7 | 2.82842712474619
         5 |       1 |                2
         5 |       1 | 2.23606797749979
....
```

**Comment by cvvergara** [comment]
Maybe not... the previous query was a user's mistake, this one is the correct one:
```
 select * from pgr_eucledianDMatrix(
'SELECT source AS id, x1 AS x, y1 AS y from edge_table WHERE source IN (1,5,6,7)
UNION
SELECT target AS id, x2 as x, y2 as y from edge_table WHERE target IN (1,5,6,7)'
) ;
 start_vid | end_vid |     agg_cost     
-----------+---------+------------------
         1 |       5 |                2
         1 |       6 | 2.23606797749979
         1 |       7 | 2.82842712474619
         5 |       1 |                2
         5 |       6 |                1
         5 |       7 |                2
         6 |       1 | 2.23606797749979
         6 |       5 |                1
         6 |       7 |                3
         7 |       1 | 2.82842712474619
         7 |       5 |                2
         7 |       6 |                3
(12 rows)
```

**Comment by cvvergara** [comment]
Its good to make a mistake
Q, what to do when the same id has different values of x,y
  - (1) That can happen because of the way the topology was created, for example, a tolerance is used when creating a topology. so a same id can have different (x,y) if data is taken from edge_table instead of edge_table_vertices_pgr.
  - OR (2) The data came from  something you might get from a web page (and you don't have control over that) and you get some data that has the same id with different (x,y). (maybe they are using a tolerance also and you still don't have control over that, and what is worse, you don't even know what that tolerance value is, or maybe they are giving wrong results, who knows, its other's people mistakes that you don't have control over)
  - OR (3) The user of the function made a mistake. (like me two comments above)
  - OR (1) & (3) happened.
  - OR (2) & (3) happened.
?????

**Comment by cvvergara** [comment]
The reason for this functions to exist is pgr_tsp V2.0, 2.1, 2.2 signatures as a preprocessing function where the results are given as a parameter to pgr_tsp.
This pgr_tsp signatures are going to be deprecated on V2.3 and for new different signatures where:
- pgr_eucledianTSP does not need a matrix
- pgr_tsp, the matrix is given as a postgres set query instead of an `array`
(see https://github.com/pgRouting/pgrouting/issues/368#issuecomment-136415210 about large arrays)

So, this functions are being deprecated together with the pgr_tsp deprecated signatures.
They will remain on the project during the 2.x versions for backwards compatibility.
